packages/generator/ni_measurement_plugin_sdk_generator/ni_measurement_plugin_client_generator/templates/measurement_plugin_client.py.mako

```diff
@@ -0,0 +1,89 @@
+<%page args="measure_docstring, configuration_metadata, output_metadata, service_class, configuration_parameters_with_type_and_values, measure_api_parameters, output_parameters_with_type, import_modules"/>\
+\
+"""Python Measurement Plug-In Client."""
+
+from functools import cached_property
+% for module, type in import_modules.items():
+% if type == "In-Built":
+${module}
+% endif
+% endfor
+from typing import List, NamedTuple
+
+import grpc
+from _helpers import (
+    _create_file_descriptor,
+    _get_measure_request,
+    _get_resolved_service,
+)
+from ni_measurement_plugin_sdk_service._internal.parameter import decoder
+from ni_measurement_plugin_sdk_service._internal.parameter.metadata import ParameterMetadata
+from ni_measurement_plugin_sdk_service._internal.stubs.ni.measurementlink.measurement.v2 import (
+    measurement_service_pb2 as v2_measurement_service_pb2,
+    measurement_service_pb2_grpc as v2_measurement_service_pb2_grpc,
+)
+% for module, type in import_modules.items():
+% if type == "Custom":
+${module}
+% endif
+% endfor
+from ni_measurement_plugin_sdk_service.discovery import DiscoveryClient
+
+<% output_type = "None" %>\
+% if output_metadata:
+
+class Output(NamedTuple):
+    """Measurement result container."""
+
+    ${output_parameters_with_type}
+<% output_type = "Output" %>\
+% endif
+
+
+class MeasurementPlugInClient:
+    """Client for accessing the Measurement Plug-In measurement services."""
+     
+    def __init__(self):
+        """Initialize the Measurement Plug-In client."""
+        self._service_class = "${service_class}"
+        self._discovery_client = DiscoveryClient()
+        self._metadata = self._measurement_service_stub.GetMetadata(v2_measurement_service_pb2.GetMetadataRequest())
+        self._configuration_metadata = ${configuration_metadata}
+        self._output_metadata = ${output_metadata}
+        _create_file_descriptor(self._metadata, self._service_class)
+        
+    @cached_property
```

Comment:
Please implement explicit locking and lazy initialization like in the other client classes.

Example: https://github.com/ni/measurement-plugin-python/blob/main/packages/service/ni_measurement_plugin_sdk_service/session_management/_client.py#L68

Reply:
Implementated with thread locking like other classes.

Reply:
Please look more closely at the other client classes, because there are some important things you missed.

1. Do lazy initialization. That means save the cached value in a member variable. Check the member variable both inside and outside of the lock.
```
        if self._stub is None:
            with self._initialization_lock:
                if self._stub is None:
                    ...
                    self._stub = ...
        return self._stub
```
2. Name the method `_get_stub`. It's a private method, so it doesn't need to uniquely identify the stub class. There is only one stub class per client class at this time.
3. The init method should take an optional discovery client and channel pool as keyword-only arguments. The point of using a channel pool is that it allows sharing channels between multiple object instances. The SessionManagementClient and PinMapClient also accept an optional grpc_channel, which is mainly for testing.
```
    def __init__(
        self,
        *,
        discovery_client: Optional[DiscoveryClient] = None,
        grpc_channel: Optional[grpc.Channel] = None,
        grpc_channel_pool: Optional[GrpcChannelPool] = None
    ) -> None:
```
4. The `_get_stub` method should lazily create a channel pool and discovery client if they were not passed to the init method.
```
        if self._stub is None:
            with self._initialization_lock:
                if self._grpc_channel_pool is None:
                    _logger.debug("Creating unshared GrpcChannelPool.")
                    self._grpc_channel_pool = GrpcChannelPool()
                if self._discovery_client is None:
                    _logger.debug("Creating unshared DiscoveryClient.")
                    self._discovery_client = DiscoveryClient(
                        grpc_channel_pool=self._grpc_channel_pool
                    )
                if self._stub is None:
```

Reply:
Sure, I have updated with the lazy initialization as suggested.